pylint/config.py

```diff
@@ -827,11 +685,465 @@ def __init__(self, *args, **kwargs):
+        # this is always going to be the global config
+        return self.global_config
+
+        path = os.path.expanduser(path)
+        path = os.path.abspath(path)
+
+        config = self._cache.get(path)
+
+        if not config:
+            config = self.global_config.copy()
+
+            parent_configs = self._get_parent_configs(path)
+            for parent_config in reversed(parent_configs):
+                config += parent_config
+
+            self._cache['path'] = config
+
+        return config
+
+    def __getitem__(self, path):
+        return self.get_config_for(path)
+
+    def __setitem__(self, path, config):
+        return self.add_config_for(path, config)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ConfigParser(object):
+    def __init__(self):
+        self._option_definitions = {}
+        self._option_groups = set()
+
+    def add_option_definitions(self, option_definitions):
+        self._option_definitions.update(option_definitions)
+
+        for _, definition_dict in option_definitions:
+            try:
+                group = optdict['group'].upper()
+            except KeyError:
+                continue
+            else:
+                self._option_groups.add(group)
+
+    def add_option_definition(self, option_definition):
+        self.add_option_definitions([option_definition])
+
+    @abc.abstractmethod
+    def parse(self, to_parse, config):
+        """Parse the given object into the config object.
+        Args:
+            to_parse (object): The object to parse.
+            config (Configuration): The config object to parse into.
+        """
+
+
+class CLIParser(ConfigParser):
+    def __init__(self, usage=''):
+        super(CLIParser, self).__init__()
+
+        self._parser = LongHelpArgumentParser(
+            usage=usage.replace("%prog", "%(prog)s"),
+            # Only set the arguments that are specified.
+            argument_default=argparse.SUPPRESS
+        )
+        # TODO: Let this be definable elsewhere
+        self._parser.add_argument('module_or_package', nargs=argparse.REMAINDER)
+
+    def add_option_definitions(self, option_definitions):
+        option_groups = collections.defaultdict(list)
+
+        for option, definition in option_definitions:
+            group, args, kwargs = self._convert_definition(option, definition)
+            option_groups[group].append(args, kwargs)
+
+        for args, kwargs in option_groups['DEFAULT']:
+            self._parser.add_argument(*args, **kwargs)
+
+        del option_groups['DEFAULT']
+
+        for group, arguments in six.iteritems(option_groups):
+            self._option_groups.add(group)
+            self._parser.add_argument_group(group.title())
+            for args, kwargs in arguments:
+                self._parser.add_argument(*args, **kwargs)
+
+    @staticmethod
+    def _convert_definition(option, definition):
+        """Convert an option definition to a set of arguments for add_argument.
```

Comment:
Let's maybe pick a style and stick with it? :) This file seems to have both Google and Sphinx style.

Reply:
Good point. Work habits sneaking in ;) I've done everything to sphinx.